pkg/controller/storageversionmigrator/resourceversion.go

```diff
@@ -0,0 +1,272 @@
+		mapper:          mapper,
+		queue:           workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), controllerName),
+	}
+
+	_, _ = svmInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			rvController.addSVM(logger, obj)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			rvController.updateSVM(logger, oldObj, newObj)
+		},
+	})
+
+	return rvController
+}
+
+func (rv *ResourceVersionController) addSVM(logger klog.Logger, obj interface{}) {
+	svm := obj.(*svmv1alpha1.StorageVersionMigration)
+	logger.V(4).Info("Adding", "svm", klog.KObj(svm))
+	rv.enqueue(svm)
+}
+
+func (rv *ResourceVersionController) updateSVM(logger klog.Logger, oldObj, newObj interface{}) {
+	oldSVM := oldObj.(*svmv1alpha1.StorageVersionMigration)
+	newSVM := newObj.(*svmv1alpha1.StorageVersionMigration)
+	logger.V(4).Info("Updating", "svm", klog.KObj(oldSVM))
+	rv.enqueue(newSVM)
+}
+
+func (rv *ResourceVersionController) enqueue(svm *svmv1alpha1.StorageVersionMigration) {
+	key, err := controller.KeyFunc(svm)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %w", svm, err))
+		return
+	}
+
+	rv.queue.Add(key)
+}
+
+func (rv *ResourceVersionController) Run(ctx context.Context) {
+	defer utilruntime.HandleCrash()
+	defer rv.queue.ShutDown()
+
+	logger := klog.FromContext(ctx)
+	logger.Info("Starting", "controller", rv.controllerName)
+	defer logger.Info("Shutting down", "controller", rv.controllerName)
+
+	if !cache.WaitForNamedCacheSync(rv.controllerName, ctx.Done(), rv.svmSynced) {
+		return
+	}
+
+	go wait.UntilWithContext(ctx, rv.worker, time.Second)
+
+	<-ctx.Done()
+}
+
+func (rv *ResourceVersionController) worker(ctx context.Context) {
+	for rv.processNext(ctx) {
+	}
+}
+
+func (rv *ResourceVersionController) processNext(ctx context.Context) bool {
+	eKey, quit := rv.queue.Get()
+	if quit {
+		return false
+	}
+	defer rv.queue.Done(eKey)
+
+	key := eKey.(string)
+	err := rv.sync(ctx, key)
+	if err == nil {
+		rv.queue.Forget(key)
+		return true
+	}
+
+	klog.FromContext(ctx).V(2).Info("Error syncing SVM resource, retrying", "svm", key, "err", err)
+	rv.queue.AddRateLimited(key)
+
+	return true
+}
+
+func (rv *ResourceVersionController) sync(ctx context.Context, key string) error {
+	logger := klog.FromContext(ctx)
+	startTime := time.Now()
+
+	// SVM is a cluster scoped resource so we don't care about the namespace
+	_, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	svm, err := rv.svmListers.Get(name)
+	if err != nil {
+		return err
+	}
+	// working with copy to avoid race condition between this and migration controller
+	toBeProcessedSVM := svm.DeepCopy()
+	gvr := getGVRFromResource(toBeProcessedSVM)
+
+	exists, err := rv.resourceExists(gvr)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		return fmt.Errorf("resource %q does not exist", gvr.String())
```

Comment:
this seems likely to produce bad outcomes.  This should instead result in a `Failed` condition, right?

Reply:
No. This check makes sure we validate if the requested resource exists "_before_" we start migration. We simply don't migrate if it doesn't exist.

Reply:
> No. This check makes sure we validate if the requested resource exists "_before_" we start migration. We simply don't migrate if it doesn't exist.

Consider the case where the GVR never exists. This would produce an unknown number of StorageMigrations failing and retrying forever.  If the resource doesn't exist, this should produce a terminal state.